### PR TITLE
31 improve best individual based termination

### PR DESCRIPTION
**Pull request by dleidreiter**

### CODE REVIEW

File: queasars/minimum_eigensolvers/base/termination_criteria.py [hunk 1]
```diff
@@ -42,34 +42,48 @@ class BestIndividualChangeTolerance(EvolvingAnsatzMinimumEigensolverBaseTerminat
...
+        self._change_history.append(change)
+        self._previous_expectation_value = population_evaluation.best_expectation_value
+
+        if len(self._change_history) < self._allowed_consecutive_violations + 1:
             return False
 
-        return True
+        max_change = max(self._change_history[-self._allowed_consecutive_violations - 1 :])
+
+        return not max_change >= self._minimum_change
```
Comment (svenpruefer) [inline queasars/minimum_eigensolvers/base/termination_criteria.py:86]: Maybe rename `max_change` to `max_change_in_last_relevant_window` or something similar? And use `return max_change_in_last_relevant_window < self._minimum_change`?

File: queasars/minimum_eigensolvers/base/termination_criteria.py [hunk 2]
```diff
@@ -42,34 +42,48 @@ class BestIndividualChangeTolerance(EvolvingAnsatzMinimumEigensolverBaseTerminat
...
 
         self._minimum_change: float = minimum_change
-        self._previous_expectation_value: float = float("inf")
+        self._allowed_consecutive_violations: int = allowed_consecutive_violations
+        self._previous_expectation_value: Optional[float] = None
+        self._change_history: list[float] = []
 
     def reset_state(self) -> None:
-        self._previous_expectation_value = float("inf")
+        self._change_history = []
```
Comment (svenpruefer) [inline queasars/minimum_eigensolvers/base/termination_criteria.py:64]: Shouldn't this also reset `self._previous_expectation_value`? It is used in line 77 for calculating the next `change`, right?

File: queasars/minimum_eigensolvers/base/termination_criteria.py [hunk 3]
```diff
@@ -80,36 +94,51 @@ class BestIndividualRelativeChangeTolerance(EvolvingAnsatzMinimumEigensolverBase
...
+        self._previous_expectation_value = population_evaluation.best_expectation_value
+        self._relative_change_history.append(relative_change)
+
+        if len(self._relative_change_history) < self._allowed_consecutive_violations + 1:
             return False
 
-        return True
+        max_change = max(self._relative_change_history[-self._allowed_consecutive_violations - 1 :])
+
+        return not max_change >= self._minimum_relative_change
```
Comment (svenpruefer) [inline queasars/minimum_eigensolvers/base/termination_criteria.py:141]: Similar as above, I would suggest to invert the inequality.

**Review by svenpruefer — APPROVED** [review]
Changelog ist missing, but the rest looks good!

File: queasars/minimum_eigensolvers/base/termination_criteria.py (hunk 1 above)
Comment (dleidreiter) [reply queasars/minimum_eigensolvers/base/termination_criteria.py]: You are right, that is much more readable. I adjusted it accordingly.

File: queasars/minimum_eigensolvers/base/termination_criteria.py (hunk 2 above)
Comment (dleidreiter) [reply queasars/minimum_eigensolvers/base/termination_criteria.py:65]: Correct. It seems i missed adding it to reset_state(). I corrected this accordingly.

File: queasars/minimum_eigensolvers/base/termination_criteria.py (hunk 3 above)
Comment (dleidreiter) [reply queasars/minimum_eigensolvers/base/termination_criteria.py]: Done :).

**Comment by dleidreiter** [comment]
I added the changelog entry.

**Pull request merged**